api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/dsd/NextAndPrevDatesCalculation.java

```diff
@@ -0,0 +1,85 @@
+package org.openmrs.module.eptsreports.reporting.calculation.dsd;
+
+import java.util.*;
+import org.openmrs.Concept;
+import org.openmrs.EncounterType;
+import org.openmrs.Location;
+import org.openmrs.Obs;
+import org.openmrs.api.context.Context;
+import org.openmrs.calculation.patient.PatientCalculationContext;
+import org.openmrs.calculation.result.CalculationResultMap;
+import org.openmrs.calculation.result.ListResult;
+import org.openmrs.module.eptsreports.metadata.HivMetadata;
+import org.openmrs.module.eptsreports.reporting.calculation.AbstractPatientCalculation;
+import org.openmrs.module.eptsreports.reporting.calculation.BooleanResult;
+import org.openmrs.module.eptsreports.reporting.calculation.common.EPTSCalculationService;
+import org.openmrs.module.eptsreports.reporting.utils.EptsCalculationUtils;
+import org.openmrs.module.reporting.common.TimeQualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+public class NextAndPrevDatesCalculation extends AbstractPatientCalculation {
+  @Override
+  public CalculationResultMap evaluate(
+      Collection<Integer> cohort,
+      Map<String, Object> parameterValues,
+      PatientCalculationContext context) {
+    EPTSCalculationService ePTSCalculationService =
+        Context.getRegisteredComponents(EPTSCalculationService.class).get(0);
+    HivMetadata hivMetadata = Context.getRegisteredComponents(HivMetadata.class).get(0);
+    Location location = (Location) context.getFromCache("location");
+    CalculationResultMap map = new CalculationResultMap();
+
+    Concept concept = (Concept) parameterValues.get("conceptId");
+    List<EncounterType> encounterTypes =
+        (List<EncounterType>) parameterValues.get("encounterTypes");
+
+    CalculationResultMap allObs =
```

Comment:
Are we able to add some conditions here to limit the number of records this function will pull for each patient?

Just wondering because this could have a performance impact

Reply:
@nthfloor  there are parameters (startDate and endDate) already passed to filter the results. This can be found in EPTSCalculationService.

Reply:
I am referring to this lookup - it will pull all the obs for all patients for a specific localtion...that could be alot of obs :)

Reply:
@nthfloor Am just checking at the requirements: **Filter all patients whose next Drugs Pick up appointment is scheduled for 83-97 days after the date of their last Drugs pick up** could we consider using the endDate to filter the obs? It will be difficult to have a start date because patients could have different last pickups that could be different from the start of the reporting period? Can that be tha case?